pkg/operator/controller/controller_dns_configmap.go

```diff
@@ -268,6 +234,50 @@ func desiredDNSConfigMap(dns *operatorv1.DNS, clusterDomain string, caBundleRevi
 	return cm, nil
 }
 
+// validateTLSSettings validates if serverName is set when tls is enabled, and validates if tls is configured only for ip addresses.
+// If a serverName is set, but tls is not enabled, then the serverName is ignored.
+func validateTLSSettings(dns *operatorv1.DNS) (*operatorv1.DNS, error) {
+	updated := dns.DeepCopy()
+	for i, server := range updated.Spec.Servers {
+		transport := server.ForwardPlugin.TransportConfig.Transport
+		tls := server.ForwardPlugin.TransportConfig.TLS
+		if transport == operatorv1.TLSTransport {
+			// tls cannot be configured without a ServerName
+			if tls == nil || tls.ServerName == "" {
+				return updated, errTransportTLSConfiguredWithoutServerName
+			}
+			// tls can only be configured for ip addresses
+			for _, upstream := range server.ForwardPlugin.Upstreams {
+				if _, _, err := net.SplitHostPort(upstream); err != nil {
+					if res := net.ParseIP(upstream); res == nil {
+						return updated, errTransportTLSConfiguredForNonIP
+					}
+				}
+			}
+		}
+		// When Transport is "" or cleartext and a ServerName is set the Corefile will ignore any other TLS settings
+		if (transport == "" || transport == operatorv1.CleartextTransport) && (tls != nil && tls.ServerName != "") {
+			logrus.Warningf("ServerName is set in %s but Transport is not set to tls. ServerName will be ignored", server.Name)
+			updated.Spec.Servers[i].ForwardPlugin.TransportConfig.TLS.ServerName = ""
+		}
+		// When Transport is "", set the default as cleartext
+		if transport == "" {
+			updated.Spec.Servers[i].ForwardPlugin.TransportConfig.Transport = operatorv1.CleartextTransport
+		}
+	}
+	transport := dns.Spec.UpstreamResolvers.TransportConfig.Transport
+	tls := dns.Spec.UpstreamResolvers.TransportConfig.TLS
+	// tls cannot be configured without a ServerName
+	if transport == operatorv1.TLSTransport && (tls == nil || tls.ServerName == "") {
+		return updated, errTransportTLSConfiguredWithoutServerName
+	}
+	// When Transport is "" or cleartext and a ServerName is set the Corefile will ignore any other TLS settings
+	if (transport == "" || transport == operatorv1.CleartextTransport) && (tls != nil && tls.ServerName != "") {
+		logrus.Warningf("ServerName is set but Transport is not set to tls. ServerName will be ignored")
+	}
```

Comment:
Should this clear `ServerName` as above?
```suggestion
	// When Transport is "", set the default as cleartext.
	if transport == "" {
		transport = operatorv1.CleartextTransport
		dns.Spec.UpstreamResolvers.TransportConfig.Transport = transport
	}
	// When Transport is "" or "Cleartext" and a ServerName is set, the Corefile will ignore any other TLS settings.
	if transport == operatorv1.CleartextTransport && tls != nil && tls.ServerName != "" {
		logrus.Warningf("ServerName is set but Transport is not set to tls. ServerName will be ignored")
		tls.ServerName = ""
	}
```
Might warrant a unit test.

Reply:
It won't be used anywhere as the transport is set to "TLS". But, it's confusing so I've changed the code to clear it as well.